include/s3select.h

```diff
@@ -1297,7 +1332,108 @@ void push_substr_from_for::builder(s3select* self, const char* a, const char* b)
   func->push_argument(expr);
 
   self->getAction()->exprQ.push_back(func);
-}  
+}
+
+void push_datediff::builder(s3select* self, const char* a, const char* b) const
+{
+  std::string token(a, b);
+
+  std::string date_op;
+
+  date_op = self->getAction()->datePartQ.back();
+  self->getAction()->datePartQ.pop_back();
+
+  std::string date_function =  "datediff_" + date_op;
+
+  __function* func = S3SELECT_NEW(self, __function, date_function.c_str(), self->getS3F());
+
+  base_statement* expr = self->getAction()->exprQ.back();
+  self->getAction()->exprQ.pop_back();
+
+  base_statement* start_position = self->getAction()->exprQ.back();
+  self->getAction()->exprQ.pop_back();
+
+  func->push_argument(start_position);
+  func->push_argument(expr);
+
+  self->getAction()->exprQ.push_back(func);
+}
+
+void push_dateadd::builder(s3select* self, const char* a, const char* b) const
+{
+  std::string token(a, b);
+
+  std::string date_op;
+
+  date_op = self->getAction()->datePartQ.back();
+  self->getAction()->datePartQ.pop_back();
+
+  std::string date_function =  "dateadd_" + date_op;
+
+  __function* func = S3SELECT_NEW(self, __function, date_function.c_str(), self->getS3F());
+
+  base_statement* expr = self->getAction()->exprQ.back();
+  self->getAction()->exprQ.pop_back();
+
+  base_statement* start_position = self->getAction()->exprQ.back();
+  self->getAction()->exprQ.pop_back();
+
+  func->push_argument(start_position);
+  func->push_argument(expr);
+
+  self->getAction()->exprQ.push_back(func);
+}
+
+void push_extract::builder(s3select* self, const char* a, const char* b) const
+{
+  std::string token(a, b);
+
+  std::string date_op;
+
+  date_op = self->getAction()->datePartQ.back();
+  self->getAction()->datePartQ.pop_back();
+
+  std::string date_function =  "extract_" + date_op;
+
+  __function* func = S3SELECT_NEW(self, __function, date_function.c_str(), self->getS3F());
+
+  base_statement* expr = self->getAction()->exprQ.back();
+  self->getAction()->exprQ.pop_back();
+
+  func->push_argument(expr);
+
+  self->getAction()->exprQ.push_back(func);
+}
+
+void push_date_part::builder(s3select* self, const char* a, const char* b) const
+{
+  std::string token(a, b);
+
+  auto date_part_op = [&](const char *s){return strncmp(a,s,strlen(s))==0;};
+
+  if(date_part_op("year"))
+  {
+    self->getAction()->datePartQ.push_back("year");
+  }else if(date_part_op("month"))
+  {
+    self->getAction()->datePartQ.push_back("month");
+  }else if(date_part_op("day"))
+  {
+    self->getAction()->datePartQ.push_back("day");
+  }else if(date_part_op("hour"))
+  {
+    self->getAction()->datePartQ.push_back("hour");
+  }else if(date_part_op("minute"))
+  {
+    self->getAction()->datePartQ.push_back("minute");
+  }else if(date_part_op("second"))
+  {
+    self->getAction()->datePartQ.push_back("second");
+  }else if(date_part_op("week"))
+  {
+    self->getAction()->datePartQ.push_back("week");
+  }
```

Comment:
no need for if-else-if ... 
the syntax-rule is accepted so 
for sure date_part_op( ... ) is one of the eligible values

Reply:
token is actually the date-part according to syntax rule.

self->getAction()->datePartQ.push_back(token;
could replace all of this if-then-else